File: leetecode/JumpGame2.cpp
#include <iostream>
#include <vector>
using namespace std;
int jump(vector<int> &nums)
{
    int n = nums.size();
    int jc = 0;
    int cj = nums[0];
    int i = 0;
    while (i < (n - 1))
    {
        int maxEl = nums[i + 1];
        int mi = i + 1;
        for (int j = i + 2; ((j < n) && (j <= (i + cj))); j++)
        {
            int temp=(nums[j] + (j - i - 1));
            // std::cout << "got val is " <<  temp<< std::endl;
            if(temp > maxEl)
            {
                maxEl = temp;
                mi = j;
            }
        }
        if (cj < n - (i + 1))
        {
            cj = nums[mi];
        }
        else
        {
            mi = cj + i;
            cj = nums[cj + i];
        }
        // std::cout << "max idnex "<< mi << " i is "<< i << " cj  is "<< cj<< std::endl;
        jc++;
        i = mi;
    }

    return jc;
}
int main()
{
    vector<int> a = {1,2,1,1,1};
    std::cout << "jup count is " << jump(a) << std::endl;
    return 0;
}

File: leetecode/JumpGame2.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        """
            Greedy: Choose that index from given range that can take us farthest ie max possible value of j 
        """
        i = 0
        j = 0
        n = len(nums)
        res = 0
        while(j<n-1):
            ci = j # default index to jump from 
            while(i<=j): # check all possible indices from i to j
                if(i+nums[i]>ci+nums[ci]): ci = i # best index to jump from
                i +=1
            j = ci + nums[ci]
            res +=1
        return res
            
